### DISCUSSION TITLE
Is there a way to map recursively?

### DISCUSSION THREAD

**Discussion by thomaslhostis** [mcarleio/konvert#94] [Q&A] [answered]
Hello,

I am trying to use Konvert so I can jump from Kapt to KSP, and I can't find a way to use it in a recursive way.

For instance, if have a class like this:

```kotlin
class Person(
    name: String,
    address: Address
)
```

And its equivalent DTO `PersonDto`.

The mapper I have to write is:
```kotlin
@Konverter
interface PersonMapper {
    fun personToPersonDto(person: Person): PersonDto
    fun addressToAddressDto(address: Address): AddressDto
}
```

But I'm not actually calling `addressToAddressDto`, it's only for Konvert to be able to map the `address` property.

It's okay when you have a few of these, I have a big tree structure and it would be nice not to have to declare a mapping function for each one of them.

Is there an option to to avoid writing a mapper for both Person and Address?

If not, what do you think about it?

Thank you for your insight and your good work :)

**Comment by mcarleio** [comment, marked as answer]
Hey, thank for for reaching out!

Currently there is no such option. `Konvert` internally builds a list of all supported `TypeConverter`s at start and simply fails with an exception if there is none matching defined when needed.

I can totally understand why this would be useful, but it would be a bigger change to Konvert which I currently have no time for.

Open topics one would have to solve along the way (if anyone [or my future self] wants to implement this):
* When not finding an appropriate `TypeConverter` a new one needs to be generated on the fly and registered to the known `TypeConverter`s
  * Should only be true for custom classes (not e.g. for `List`, `Int` or `String`)
* For not repeating to generate the same code, it should be reusable, but:
  * Where should the generated code be generated?
  * With what visibility?
  * With which configuration (`@Konfig`)?
* There should be an option so that you can enable/disable the behavior
* Should those generated functions be part of the current `META-INF` files (or however #78 will be solved)?
